pkg/client/block/client_test.go

```diff
@@ -0,0 +1,125 @@
+	"pocket/internal/testclient/testeventsquery"
+	"pocket/pkg/client"
+	"pocket/pkg/client/block"
+	eventsquery "pocket/pkg/client/events_query"
+)
+
+const blockAssertionLoopTimeout = 100 * time.Millisecond
+
+func main() {
+	fmt.Println("HELLOO!!!")
+}
+
+func TestBlockClient(t *testing.T) {
+	var (
+		expectedHeight     = int64(1)
+		expectedHash       = []byte("test_hash")
+		expectedBlockEvent = &testBlockEvent{
+			Block: comettypes.Block{
+				Header: comettypes.Header{
+					Height: 1,
+					Time:   time.Now(),
+					LastBlockID: comettypes.BlockID{
+						Hash: expectedHash,
+					},
+				},
+			},
+		}
+		ctx = context.Background()
+	)
+
+	// Set up a mock connection and dialer which are expected to be used once.
+	connMock, dialerMock := testeventsquery.OneTimeMockConnAndDialer(t)
+	connMock.EXPECT().Send(gomock.Any()).Return(nil).Times(1)
+	// Mock the Receive method to return the expected block event.
+	connMock.EXPECT().Receive().DoAndReturn(func() ([]byte, error) {
+		blockEventJson, err := json.Marshal(expectedBlockEvent)
+		require.NoError(t, err)
+		return blockEventJson, nil
+	}).AnyTimes()
+
+	// Set up events query client dependency.
+	dialerOpt := eventsquery.WithDialer(dialerMock)
+	eventsQueryClient := testeventsquery.NewLocalnetClient(t, dialerOpt)
+	deps := depinject.Supply(eventsQueryClient)
+
+	// Set up block client.
+	blockClient, err := block.NewBlockClient(ctx, deps, testclient.CometWebsocketURL)
+	require.NoError(t, err)
+	require.NotNil(t, blockClient)
+
+	// Run LatestBlock and CommittedBlockSequence concurrently because they can
+	// block, leading to an unresponsive test. This function sends multiple values
+	// on the actualBlockCh which are all asserted against in blockAssertionLoop.
+	// If any of the methods under test hang, the test will time out.
+	var (
+		actualBlockCh = make(chan client.Block, 1)
+		done          = make(chan struct{}, 1)
+	)
+	go func() {
+		// Test LatestBlock method.
+		actualBlock := blockClient.LatestBlock(ctx)
+		require.Equal(t, expectedHeight, actualBlock.Height())
+		require.Equal(t, expectedHash, actualBlock.Hash())
+
+		// Test CommittedBlockSequence method.
+		blockObservable := blockClient.CommittedBlocksSequence(ctx)
+		require.NotNil(t, blockObservable)
+
+		// Ensure that the observable is replayable via Last.
+		actualBlockCh <- blockObservable.Last(ctx, 1)[0]
+
+		// Ensure that the observable is replayable via Subscribe.
+		blockObserver := blockObservable.Subscribe(ctx)
+		for block := range blockObserver.Ch() {
+			actualBlockCh <- block
+			break
+		}
+
+		// Signal test completion
+		done <- struct{}{}
+	}()
+
+blockAssertionLoop:
+	for {
+		select {
+		case actualBlock := <-actualBlockCh:
+			require.Equal(t, expectedHeight, actualBlock.Height())
+			require.Equal(t, expectedHash, actualBlock.Hash())
+		case <-done:
+			break blockAssertionLoop
+		case <-time.After(blockAssertionLoopTimeout):
+			t.Fatal("timed out waiting for block event")
+		}
```

Comment:
Can you help me understand what happens here.

1. We block on all the cases
2. The first one geta a response (twice)
3. The 2nd one get a response (once)
4. Success

Is that it?

Reply:
:heavy_check_mark:

Reply:
#PUC